fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessageHomescreenTest.kt

```diff
@@ -0,0 +1,119 @@
+import org.mozilla.fenix.nimbus.Messaging
+import org.mozilla.fenix.nimbus.StyleData
+import org.mozilla.fenix.ui.robots.homeScreen
+
+class NimbusMessageHomescreenTest {
+    private lateinit var mDevice: UiDevice
+    private lateinit var mockWebServer: MockWebServer
+
+    private var messageButtonLabel = "CLICK ME"
+    private var messageText = "Some Nimbus Messaging text"
+    private var messageTitle = "A Nimbus title"
+
+    @get:Rule
+    val homeActivityTestRule = HomeActivityIntentTestRule.withDefaultSettingsOverrides(
+        skipOnboarding = true,
+    ).withIntent(
+        Intent().apply {
+            action = Intent.ACTION_VIEW
+        },
+    )
+
+    @Rule
+    @JvmField
+    val retryTestRule = RetryTestRule(3)
+
+    @Before
+    fun setUp() {
+        // Set up nimbus message
+        FxNimbus.features.messaging.withInitializer {
+            // FML generated objects.
+            Messaging(
+                messages = mapOf(
+                    "test-message" to MessageData(
+                        action = Res.string("TEST ACTION"),
+                        style = "URGENT",
+                        buttonLabel = Res.string(messageButtonLabel),
+                        text = Res.string(messageText),
+                        title = Res.string(messageTitle),
+                        trigger = listOf("ALWAYS"),
+                    ),
+                ),
+                styles = mapOf(
+                    "TEST STYLE" to StyleData(),
+                ),
+                actions = mapOf(
+                    "TEST ACTION" to "https://example.com",
+                ),
+                triggers = mapOf(
+                    "ALWAYS" to "true",
+                ),
+            )
+        }
+
+        // Remove some homescreen features not needed for testing
+        FxNimbus.features.homescreen.withInitializer {
+            // These are FML generated objects and enums
+            Homescreen(
+                sectionsEnabled = mapOf(
+                    HomeScreenSection.JUMP_BACK_IN to false,
+                    HomeScreenSection.POCKET to false,
+                    HomeScreenSection.POCKET_SPONSORED_STORIES to false,
+                    HomeScreenSection.RECENT_EXPLORATIONS to false,
+                    HomeScreenSection.RECENTLY_SAVED to false,
+                    HomeScreenSection.TOP_SITES to false,
+                ),
+            )
+        }
+
+        mDevice = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
+        mockWebServer = MockWebServer().apply {
+            dispatcher = AndroidAssetDispatcher()
+            start()
+        }
+        // refresh message store
+        appContext.components.appStore.dispatch(
+            AppAction.MessagingAction.Restore,
+        )
+        appContext.components.appStore.dispatch(
+            AppAction.MessagingAction.Evaluate(
+                MessageSurfaceId.HOMESCREEN,
+            ),
+        )
```

Comment:
I tried to remove these, but could no longer see the message.

Is this:
- a bug in the app?
- necessary due to a limitation in the testing method?

I _think_ it might be that [`restoreMessaging`](https://github.com/mozilla-mobile/firefox-android/blob/2c38a8682df8470da543faafa9725480d558587c/fenix/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#L494) is called from `onCreate`.

If that's the case, then I think re-writing this by calling `restoreMessaging` from the test, or moving `restoreMessaging` from the `FenixApplication` to `HomeActivity`'s `onCreate` method, so this test can treat the app as a black box.

Reply:
So @jonalmeida is the most familiar with this, as he helped me a good bit to get this figured out. This is a bit out of my wheelbase, what do you think Jon? I think moving the ```restoreMessaging``` call to ```HomeActivity``` makes sense, but not sure what limitations that will bring.

Reply:
Good catch! The `Evaluate` is necessary and was a leftover artifact from our debugging, so we should remove that.

The `Restore` is needed and I tried the patch below to call `restoreMessaging` which works for the test. 


<details>

<summary>Patch</summary>

```diff
diff --git a/fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessageHomescreenTest.kt b/fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessageHomescreenTest.kt
index 6759fe81e4..fc9ca2268f 100644
--- a/fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessageHomescreenTest.kt
+++ b/fenix/app/src/androidTest/java/org/mozilla/fenix/ui/NimbusMessageHomescreenTest.kt
@@ -9,17 +9,14 @@ import org.junit.Before
 import org.junit.Rule
 import org.junit.Test
 import org.mozilla.experiments.nimbus.Res
-import org.mozilla.fenix.components.appstate.AppAction
-import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.FenixApplication
 import org.mozilla.fenix.helpers.AndroidAssetDispatcher
 import org.mozilla.fenix.helpers.HomeActivityIntentTestRule
 import org.mozilla.fenix.helpers.RetryTestRule
-import org.mozilla.fenix.helpers.TestHelper.appContext
 import org.mozilla.fenix.nimbus.FxNimbus
 import org.mozilla.fenix.nimbus.HomeScreenSection
 import org.mozilla.fenix.nimbus.Homescreen
 import org.mozilla.fenix.nimbus.MessageData
-import org.mozilla.fenix.nimbus.MessageSurfaceId
 import org.mozilla.fenix.nimbus.Messaging
 import org.mozilla.fenix.nimbus.StyleData
 import org.mozilla.fenix.ui.robots.homeScreen
@@ -94,14 +91,8 @@ class NimbusMessageHomescreenTest {
             start()
         }
         // refresh message store
-        appContext.components.appStore.dispatch(
-            AppAction.MessagingAction.Restore,
-        )
-        appContext.components.appStore.dispatch(
-            AppAction.MessagingAction.Evaluate(
-                MessageSurfaceId.HOMESCREEN,
-            ),
-        )
+        val application = (homeActivityTestRule.activity.application as FenixApplication)
+        application.restoreMessaging()
     }
 
     @After
diff --git a/fenix/app/src/main/java/org/mozilla/fenix/FenixApplication.kt b/fenix/app/src/main/java/org/mozilla/fenix/FenixApplication.kt
index 4ff7925530..718f60d9b9 100644
--- a/fenix/app/src/main/java/org/mozilla/fenix/FenixApplication.kt
+++ b/fenix/app/src/main/java/org/mozilla/fenix/FenixApplication.kt
@@ -491,7 +491,8 @@ open class FenixApplication : LocaleAwareApplication(), Provider {
         }
     }
 
-    private fun restoreMessaging() {
+    @VisibleForTesting
+    internal fun restoreMessaging() {
         if (settings().isExperimentationEnabled) {
             components.appStore.dispatch(AppAction.MessagingAction.Restore)
         }

```

</details>

Reply:
Could you explain this a bit for me?

```val application = (homeActivityTestRule.activity.application as FenixApplication)```

This looks like a generator of sorts?

Reply:
`homeActivityTestRule` is a UI test helper, you can have a look at it's implementation for all the things it can do. It's mostly useful here is so we can access the `HomeActivity` which will have access to the `Application` class as that is the main app class that starts all the UI ones. 

What we get back is just an `Application` class so we cast it to our specific type `FenixApplication` since we know that this will always be the case. From there, we can use the method `restoreMessaging` in there, which will then dispatch the correct action.

Let me know if poorly explained anything. The question was a bit broad. :)

Reply:
Great thanks, I am trying to learn more about kotlin as I write these tests, thanks for that explanation